pkg/reconciler/controllers/remediator.go

```diff
@@ -0,0 +1,348 @@
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/metrics"
+	"kpt.dev/configsync/pkg/remediator"
+	"kpt.dev/configsync/pkg/remediator/cache"
+	"kpt.dev/configsync/pkg/status"
+	syncerclient "kpt.dev/configsync/pkg/syncer/client"
+	syncerreconcile "kpt.dev/configsync/pkg/syncer/reconcile"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Remediator is a controller that watches for timer-based events.
+// When an object is added to the remediate resource list, it enqueues the GKNN
+// of the object as a remediate request. The reconciler dequeues the request
+// and remediate the object to match the declared resource.
+type Remediator struct {
+	// ControllerCtx is set in reconciler/main.go and is canceled by the Finalizer.
+	ControllerCtx context.Context
+	// DoneCh indicates the Remediator stops processing new requests.
+	// It is a signal for the Finalizer to continue to finalize managed resources.
+	DoneCh chan struct{}
+	// SyncScope is the scope of the reconciler, either a namespace or ':root'
+	SyncScope declared.Scope
+	// SyncName is the name of the RootSync or RepoSync.
+	SyncName string
+	// Applier is the client to correct drift.
+	syncerreconcile.Applier
+	// declared.Resources is the cache of the declared resource list. It is used
+	// as the source of truth for drift correction.
+	*declared.Resources
+	// cache.RemediateResources is the cache of the remediate resource list. It is
+	// used to get the actual Kubernetes object for a particular GKNN.
+	*cache.RemediateResources
+	// Remediator is the actor to check whether it is ready to start reconciliation.
+	*remediator.Remediator
+}
+
+// Reconcile processes the remediate request to correct drift.
+func (r *Remediator) Reconcile(_ context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if errors.Is(r.ControllerCtx.Err(), context.Canceled) {
+		// Drain the request queue
+		klog.V(3).Infof("Stop remediating new objects as the context is canceled..")
+		return ctrl.Result{}, nil
+	}
+
+	klog.V(1).Infof("New remediator reconciliation for %q", req.Name)
+
+	id, err := core.FromID(req.Name)
+	if err != nil {
+		klog.Errorf("invalid request: %v", err)
+		return ctrl.Result{}, nil
+	}
+
+	// marks the Remediator is working to block the Applier if it can remediate.
+	if !r.StartRemediating() {
+		// Mark the object as done so that it can be requeued later when the Applier is not running.
+		klog.V(1).Infof("Skip remediating object %s because the Remediator is paused", id)
+		r.DoneObj(id)
+		return ctrl.Result{}, nil
+	}
+	// Mark the Remediator as done before it exits, so the Syncer can start the apply work.
+	defer r.DoneRemediating()
+
+	obj, found := r.GetObjs()[id]
+	if !found {
+		klog.V(4).Infof("Skip remediating object %q because it is no longer on the remediate list", id)
```

Comment:
What does "no longer on the remediate list" mean in this case? The watch saw an event but we dont care about that object for some reason?